README
Welcome to Gutenprint 5.3.4!  Please read these notes carefully.

*** NOTE TO PACKAGERS: Please read the file README.package for issues
    that are of interest to distributors and packagers of Gutenprint.
    It is not necessary for end users of Gutenprint to read this file.

================================================================

I) INTRODUCTION

    Gutenprint is a suite of printer drivers for UNIX, Linux, and
    Macintosh OS X systems (10.6 through currently 10.14) that use
    CUPS as their printing system.  Gutenprint currently supports over
    3000 printers.  It also includes an enhanced Print plug-in for
    GIMP that replaces the print plug-in packaged with the GIMP
    distribution.

    Gutenprint was renamed from its original name of Gimp-Print in
    order to clearly distinguish it from GIMP.  While this package
    started out as the Print plugin for GIMP, it has expanded into a
    collection of general purpose printer drivers, and the enhanced
    Print plugin for GIMP is now only a small part of the package.
    Furthermore, the name Gutenprint recognizes Johannes Gutenberg,
    the inventor of the movable type printing press.  Finally, the
    word "guten" means "good" in German.

    Please read the user's manual (doc/gutenprint-users-manual.pdf)
    carefully, as it covers much more information than this README
    file.  In addition, please read the release notes (NEWS), as it
    has information on the latest changes.

    Gutenprint supports only the printer part of multi-function
    devices (devices that typically include scanning, copying, and fax
    capabilities).

    We recommend that all users who wish to use this package for
    general purpose printing use CUPS.  CUPS provides an excellent
    web-based interface for easy printer administration, and all
    CUPS-enabled applications (OpenOffice.org, all KDE apps, etc.) and
    printing dialogs (kprinter, xpp, etc.) show all the options in
    easy to use GUIs.  Please visit http://www.cups.org/ for
    information on downloading and installing CUPS.


================================================================

II) INSTALLATION

    This section is provided for people installing the package from
    source.  If you are installing from a pre-compiled package (such
    as the Macintosh OS X package, or a package provided by your
    distribution vendor), you may skip this section.

    Installing packages from source requires some level of system
    administration skills along with superuser privileges.  Superuser
    privilege allows you to perform actions that may be damaging to
    your system.  If you are not comfortable with the material
    discussed here and in the release notes, or with performing simple
    administrative actions, we recommend that you not attempt to
    install this package from source.

    If you are building Gutenprint from the git repository, you need
    to take some additional steps.  Please read doc/README.maintaining
    for instructions on how to build directly from git.

    Before beginning, please read the release notes carefully for any
    updates.

----------------

A) OVERVIEW

    Gutenprint includes the following primary components:

      - The core driver library (required for everything else)
      - A CUPS (Common UNIX Printing System) driver
      - A CUPS backend, required for selected dye-sublimation printers
      - A command-line utility to administer and maintain Epson
        printers
      - An enhanced Print plug-in for GIMP
      - GTK-based UI libraries (used by the enhanced GIMP plug-in)

    This package requires the use of GNU Make to compile.  On systems
    with both GNU make and another make installed, GNU make may be
    named `gmake' or `gnumake'.  BSD users in particular must take
    care to use 'gmake'.

    The general procedure to build Gutenprint is as follows:

    ./configure [options]
    make
    make install

    The "configure" script, which must be run before doing anything
    else, controls which of these components are built along with any
    options desired.  This script determines what software is
    installed on your system and what components will be compatible
    with it, and prepares the package to be built and installed.  If
    the script emits any warnings or errors, please be certain that
    you understand them before proceeding.

    After the configure script has completed its work, it will print
    a summary of its choices.  Please ensure that it is correct before
    proceeding.  The summary looks like this:

	Configuration Summary:
	------------- --------

	If you have any problems, please report the information below to
	gimp-print-devel@lists.sourceforge.net

	================================================================
	  Release: gutenprint 5.3.3 generated on 25 Aug 2019
	  Generated at Sun Aug 25 11:11:14 EDT 2019 by rlk

	  Features:
	    Build CUPS:                                 yes, installing in /usr
		Build CUPS 1.2 enhancements:            yes
		Build CUPS PPD files:                   no
		Generate PS level 3 CUPS PPD files:     yes
		Build genppd statically:                yes
		Build CUPS dyesub USB backend:          yes
	    Build EPSON inkjet utility:                 yes
	    Build enhanced Print plugin for GIMP:       yes
		GIMP plugin will be named:              gutenprint
		Install plugin(s) in home directory:    no
	    Build test programs:                        yes
	    Build testpattern generator:                yes

	  Installation summary:
	    Installation prefix:                        /usr/local
	    Exec prefix:                                /usr/local (${prefix})
	    Data directory:                             /usr/local/share/gutenprint
	    Library directory:                          /usr/local/lib64/gutenprint (${exec_prefix}/lib64/gutenprint)
	    Executable directory:                       /usr/local/bin (/usr/local/bin)
	    XML data directory:                         /usr/local/share/gutenprint/5.3/xml
	    Module directory:                           /usr/local/lib64/gutenprint/5.3/modules (${exec_prefix}/lib64/gutenprint/5.3/modules)
	    Install sample images:                      yes

	  General configuration:
	    Configure arguments:                        --enable-maintainer-mode
	    Compiler:                                   gcc
	    Compiler Version:                           gcc version 7.4.1 20190725 [gcc-7-branch revision 273795] (SUSE Linux)
	    Compiler options:                           -Disfinite=finite  -O3  -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -Wwrite-strings -Werror-implicit-function-declaration -Winline -Wformat=2 -finline-limit=131072 -Wformat -Werror=format-security -Wmissing-prototypes -D_POSIX_C_SOURCE=200809L -std=c99 -pedantic -Waggregate-return -Wcast-qual -Wshadow -Wredundant-decls
	    Build static libraries:                     yes
	    Build shared libraries:                     yes
	    Maintainer mode:                            yes
	    Use i18n:                                   yes
	    Generate profiling information:             no
	    Generate debugging symbols:                 no
	    Use modules:                                static
	    Use readline libraries:                     yes, extra arguments: -lncurses
	    uname -a output:                            Linux dsl092-065-009 4.12.14-lp151.28.13-default #1 SMP Wed Aug 7 07:20:16 UTC 2019 (0c09ad2) x86_64 x86_64 x86_64 GNU/Linux
	================================================================


    The CUPS and GIMP components require that the appropriate packages
    be installed on your system.  Many distribution vendors separate
    packages into runtime and development packages.  The development
    packages are required in order to successfully compile Gutenprint.
    In many cases the configure script cannot detect whether these
    development packages are installed, in which case you will get an
    error during compilation.  Also note that it is necessary for the
    versions of the development packages to match exactly the versions
    of the runtime packages they are related to.

    PLEASE INSPECT THE OUTPUT CAREFULLY BEFORE PROCEEDING.  If you
    expect to build a particular component (particularly the CUPS
    driver, CUPS backends, or enhanced Print plugin for GIMP), and
    the configure output states that that component will not be built,
    please make certain that you have the required development packages
    installed, as described below in the sections for the particular
    components.


----------------

B) CORE DRIVER LIBRARY

    The core driver library, which is always built, contains all of
    the printer drivers comprising the Gutenprint package.  When the
    package is built, it consists of a core library (libgutenprint), a
    set of XML files, and depending upon the options selected,
    additional libraries containing the drivers for each family of
    printers.

    By default dynamically loadable modules (plug-ins) will be built
    for the family drivers, and loaded at run-time if your operating
    system supports it.  If you experience problems,
    --with-modules=dlopen or --with-modules=ltdl may be used to select
    the module loading method (dlopen is the default, but GNU libltdl
    is more portable), or --with-modules=static or --without-modules
    disables them.  If you wish to compile Gutenprint entirely
    statically (with no use of shared libraries), you may use the
    option --disable-shared.


----------------

C) CUPS

    Gutenprint is normally used as a driver under CUPS (Common UNIX
    Printing System).  Full description of CUPS is beyond the scope of
    this README file; full information may be found at
    http://www.cups.org.  Gutenprint 5.3 requires CUPS 1.1.9 or
    higher.  We recommend use of 1.1.15 or above; that release of CUPS
    fixes some important bugs.  Gutenprint supports CUPS 1.2, and
    supports the CUPS 1.2 enhancements, including dynamic PPD file
    generation.

    Gutenprint will fail to configure if the necessary CUPS
    development packages are not installed.  This is to protect users
    against attempting to install newer versions of Gutenprint without
    installing the CUPS development package.  This results in an
    installation that does not work with CUPS, which has created
    confusion in the past.

    Normally, you should install the CUPS development package, which
    on most systems is "cups-devel", "cups-dev", or similar, using
    your operating system's package manager.

    If you really want to install Gutenprint without CUPS, you may run
    configure with "--without-cups".  This may be useful if you intend
    to use only the Gutenprint plugin for GIMP, or you intend to use
    the Gutenprint library to support a custom printing solution that
    does not involve CUPS.  However, for most users, simply running
    configure --without-cups is not the correct course of action.

    If Gutenprint thinks it has detected a correct installation of
    CUPS, but that installation is in fact not correct, you will get
    errors when you compile the package.  The errors will typically
    look like this; the important error is "cups/cups.h: No such file
    or directory" (other ".h" files may also be missing, depending
    upon the missing packages):

	make[3]: Entering directory `/opt/gutenprint/gutenprint-5.3.0-pre2/src/cups'
	gcc -DHAVE_CONFIG_H -I. -I../..  -I../../include -I../../include
	   -Wall -Wcast-align -Wstrict-prototypes -Wmissing-prototypes
	   -Wmissing-declarations -Wnested-externs -Wwrite-strings
	   -Werror-implicit-function-declaration -Winline -Wformat=2
	   -finline-limit=131072  -DCUPS_DRIVER_INTERFACE -DALL_LINGUAS='"cs da
	   de el en_GB es fr hu it ja nb nl pl pt sk sv zh_TW"'
	   -Disfinite=finite  -O6 -MT gutenprint_5_3-genppd.o -MD -MP -MF
	   .deps/gutenprint_5_3-genppd.Tpo -c -o gutenprint_5_3-genppd.o
	   `test -f 'genppd.c' || echo './'`genppd.c
	genppd.c:79:23: error: cups/cups.h: No such file or directory
	genppd.c:80:25: error: cups/raster.h: No such file or directory
	make[3]: *** [gutenprint_5_3-genppd.o] Error 1
	make[3]: Leaving directory `/opt/gutenprint/gutenprint-5.3.0-pre2/src/cups'
	make[2]: *** [all-recursive] Error 1
	make[2]: Leaving directory `/opt/gutenprint/gutenprint-5.3.0-pre2/src'
	make[1]: *** [all-recursive] Error 1
	make[1]: Leaving directory `/opt/gutenprint/gutenprint-5.3.0-pre2'
	make: *** [all] Error 2

    Installing the CUPS driver for Gutenprint 5.3 will not interfere
    with your ability to continue using the Gutenprint 5.2 or earlier
    drivers.

    WARNING: The "canon" and "epson" CUPS backends that previously
    were distributed with Gutenprint are no longer distributed, so you
    will need to ensure that none of your printer queues use these

NEWS
Welcome to Gutenprint 5.3.4!  Please read these release notes
carefully.  Gutenprint 5.2 is still considered the stable release
line.

Gutenprint is a suite of printer drivers for UNIX, Linux, and
Macintosh OS X systems (10.6 and above) that use CUPS as their
printing system.  Gutenprint currently supports over 3000 printers.
It also includes an enhanced Print plug-in for GIMP that replaces the
print plug-in packaged with the GIMP distribution.

*** NOTE TO PACKAGERS: Please read the file README.package for issues
    that are of interest to distributors and packagers of Gutenprint.
    It is not necessary for end users of Gutenprint to read this file.

These release notes contain the following sections:

I)    General Requirements
II)   Changes from Previous Releases
III)  Exceptions and Workarounds
	A) General Issues
	B) Printer-Specific Issues
	C) Build/Installation Issues
IV)   Overall changes from Gutenprint 5.2
	A) General User-visible Changes
	B) New Functionality
	C) Changes to the Enhanced Print plugin for GIMP
	D) Changes to the CUPS interface
	E) Architectural Changes
V)    Critical Upgrade Note from Gutenprint 5.0 or Earlier

================================================================

I) GENERAL REQUIREMENTS

Gutenprint will run on any reasonably modern computer running Linux,
Macintosh OS X (10.6 through currently 10.14), Solaris, or any other
UNIX-like operating system.  If you plan to compile this package from
source, you will also need an ANSI C compiler, such as gcc
(recommended).  A compiler is not required if you are installing a
pre-compiled package.

Processor and memory requirements vary depending upon the printer and
runtime options selected; any system dating from 2005 or later should
easily suffice.  We suggest that you have at least 64 MB of memory for
general purpose printing, 256 MB or more for high quality printing on
a good printer, and 1 GB or more for large format printing at high
resolution.  You should have at least 50 MB of free disk space to
compile and install Gutenprint.  Disk space requirements for printing
will vary depending upon how you use Gutenprint, but are generally
modest except as noted below.  We recommend a processor speed of at
least 300 MHz.  Fast printers may require a faster processor to
achieve maximum printing speed.

For general use, you should have the Common UNIX Printing System, CUPS
(version 1.1.15 or above).  Please read the rest of the release notes,
in particular the Exceptions and Workarounds, for full details on
installation, as there is important information to be aware of.  CUPS
is the printing system used on Macintosh OS X 10.3 and above, and many
other systems use it.  The combination of CUPS and Gutenprint provides
a flexible, general purpose printing system capable of producing the
highest quality output with any of the printers supported by this
package.  We strongly recommend using CUPS with Gutenprint as a
general-purpose printing solution.

The enhanced Print plug-in for GIMP requires GIMP 2.0 or above (GIMP
2.2 recommended).  This plug-in will work with any printing system,
and offers a comprehensive user interface to control all aspects of
the printing process.  If you are printing photographs in large format
from GIMP at very high resolution, disk space requirements may be
substantial, and we recommend at least 2 GB of free disk space for
that purpose.

Users of Macintosh 10.6 (Snow Leopard) and above can use this package.
For ease of installation, a pre-built package with installer is
normally supplied a few days after the release of the source package.
We strongly recommend that OS X users use the pre-built package rather
than attempt to build it themselves.  Packages are not always supplied
for development releases.

The README file included with this package provides full instructions
for building and installing Gutenprint.
================================================================


II) MAJOR CHANGES FROM PREVIOUS RELEASES

* Changes from Gutenprint 5.3.3

  1) Gutenprint (any release) does not currently work on MacOS 10.15
     (Catalina).  We do not as of this time have a fix in hand; if you
     want to use Gutenprint, you should stick with 10.14 (Mojave) or
     below.

  2) A long-standing gamma correction bug affecting all dye-sublimation
     printers has been corrected, resulting in visibly improved contrast,
     tonal range, and color fidelity.  However, users upgrading from older
     versions can continue using the old gamma behaviour, by setting the
     "StpLegacyDyesubGamma" option to "true"

  3) Support for the following dye sublimation & thermal printers has been added:

     Canon SELPHY CP1500
     Citizen CX-02W
     Citizen CZ-01
     DNP DS480
     DNP DS680
     DNP QW410
     HiTi P461 Prinhome
     HiTi P510K / P510L / P510S / P510SI
     HiTi P518A / P518S (EXPERIMENTAL)
     HiTi P520L
     HiTi P525L
     HiTi P720L / P728L (EXPERIMENTAL)
     HiTi P750L (EXPERIMENTAL)
     Kodak 6900 (EXPERIMENTAL)
     Mitsubishi CP30D series
     Mitsubishi CP9500-S / CP9600-S / CP9820-S
     Mitsubishi CP-D70DW-S
     Mitsubishi CP-M1 & CP-M15 series
     Mitsubishi CP-W5000 series
     Nidec Copal DPB-7000
     Olmec OP1000
     Sony UP-CX1
     Sont UP-D711MD
     Sony UP-971AD
     Sony UP-991AD (EXPERIMENTAL)
     Stryker SDP1000

  4) The following printers have received siginificant bugfixes or
     major enhancements:

     Canon SELPHY CP790
     DNP DS620
     DNP DS820
     Fujifilm ASK-2500
     Kodak 605
     Kodak 6800/6850
     Kodak 7000 series
     Kodak 8800 / 9810
     Kodak 8810
     Mitsubishi CP-D90 series
     Mitsubishi CP9800 series
     Sinfonia CHC-S2245/S3
     Sony UP-D898 family
     Sony UP-DR80MD
     Sony UP-DR200

     Note that for the Mitsubishi CP-D90 series, when upgrading from a
     prior release, you will need to remove and re-create your printer
     queue for printing to function properly.


src/gimp2/print.c
/*
 *
 *   Print plug-in for the GIMP.
 *
 *   Copyright 1997-2000 Michael Sweet (mike@easysw.com) and
 *	Robert Krawitz (rlk@alum.mit.edu)
 *
 *   This program is free software; you can redistribute it and/or modify it
 *   under the terms of the GNU General Public License as published by the Free
 *   Software Foundation; either version 2 of the License, or (at your option)
 *   any later version.
 *
 *   This program is distributed in the hope that it will be useful, but
 *   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *   for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <gutenprintui2/gutenprintui.h>
#include "print_gimp.h"

#include <sys/types.h>
#include <signal.h>
#include <ctype.h>
#include <sys/wait.h>

#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "print-intl.h"

/*
 * Local functions...
 */

static void    query           (void);
static void    run             (const char        *name,
                                gint               nparams,
                                const GimpParam   *param,
                                gint              *nreturn_vals,
                                GimpParam        **return_vals);
static int     do_print_dialog (const gchar *proc_name,
                                gint32       image_ID);

/*
 * Globals...
 */

GimpPlugInInfo	PLUG_IN_INFO =		/* Plug-in information */
{
  NULL,  /* init_proc  */
  NULL,  /* quit_proc  */
  query, /* query_proc */
  run,   /* run_proc   */
};

static stpui_plist_t gimp_vars;

/*
 * 'main()' - Main entry - just call gimp_main()...
 */

MAIN()

/*
 * 'query()' - Respond to a plug-in query...
 */

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wcast-qual"
static void
query (void)
{
  static GimpParamDef args[] =
  {
    { GIMP_PDB_INT32,	  (BAD_CONST_CHAR) "run_mode",(BAD_CONST_CHAR) "Interactive, non-interactive" },
    { GIMP_PDB_IMAGE,	  (BAD_CONST_CHAR) "image",   (BAD_CONST_CHAR) "Input image" },
    { GIMP_PDB_DRAWABLE,(BAD_CONST_CHAR) "drawable",(BAD_CONST_CHAR) "Input drawable" },
    { GIMP_PDB_STRING,	(BAD_CONST_CHAR) "driver",	(BAD_CONST_CHAR) "Printer driver short name" },
    { GIMP_PDB_STRING,	(BAD_CONST_CHAR) "printer_queue",(BAD_CONST_CHAR) "CUPS Printer Queue" },
    { GIMP_PDB_FLOAT,   (BAD_CONST_CHAR) "left",    (BAD_CONST_CHAR) "Left offset (points, -1 = centered)" },
    { GIMP_PDB_FLOAT,   (BAD_CONST_CHAR) "top",     (BAD_CONST_CHAR) "Top offset (points, -1 = centered)" },
    { GIMP_PDB_INT32,   (BAD_CONST_CHAR) "length_key_value_array", (BAD_CONST_CHAR) "Length of the key-value array" },
    { GIMP_PDB_STRINGARRAY, (BAD_CONST_CHAR) "keys", (BAD_CONST_CHAR) "Key-value pairs for Gutenprint Settings" },
  };

  static const gchar *blurb = "This plug-in prints images from The GIMP using Gutenprint directly.";
  static const gchar *help  = "Prints images to many printers.";
  static const gchar *auth  = "Michael Sweet <mike@easysw.com> and Robert Krawitz <rlk@alum.mit.edu>";
  static const gchar *copy  = "Copyright 1997-2006 by Michael Sweet and Robert Krawitz";
  static const gchar *types = "RGB*,GRAY*,INDEXED*";

  gimp_plugin_domain_register (cast_safe(PACKAGE), cast_safe(PACKAGE_LOCALE_DIR));

  do_gimp_install_procedure(blurb, help, auth, copy, types, G_N_ELEMENTS(args),
			    args);
}
#pragma GCC diagnostic pop

static guchar *gimp_thumbnail_data = NULL;

static guchar *
stpui_get_thumbnail_data_function(void *image_ID, gint *width, gint *height,
				  gint *bpp, gint page)
{
  if (gimp_thumbnail_data)
    g_free(gimp_thumbnail_data);
  gint x = gimp_image_width(p2gint(image_ID));
  gint y = gimp_image_height(p2gint(image_ID));

  /* GIMP limits thumbnails to 1024px! */
  if (*width > 1024)
    *width = 1024;
  if (*height > 1024)
    *height = 1024;

  if (*width > x)
    *width = x;
  if (*height > y)
    *height = y;
  gimp_thumbnail_data =
    gimp_image_get_thumbnail_data(p2gint(image_ID), width, height, bpp);
  return gimp_thumbnail_data;
}

/*
 * 'run()' - Run the plug-in...
 */

volatile int SDEBUG = 1;

static void
run (const char        *name,		/* I - Name of print program. */
     gint               nparams,	/* I - Number of parameters passed in */
     const GimpParam   *param,		/* I - Parameter values */
     gint              *nreturn_vals,	/* O - Number of return values */
     GimpParam        **return_vals)	/* O - Return values */
{
  GimpDrawable	*drawable;	/* Drawable for image */
  GimpRunMode	 run_mode;	/* Current run mode */
  GimpParam	*values;	/* Return values */
  gint32         drawable_ID;   /* drawable ID */
  GimpExportReturn export = GIMP_EXPORT_CANCEL;    /* return value of gimp_export_image() */
  gdouble xres, yres;
  char *image_filename;
  stpui_image_t *image;
  gint32 image_ID;
  gint32 base_type;
  stp_parameter_t desc;
  if (getenv("STP_DEBUG_STARTUP"))
    {
      fprintf(stderr, "pid is %d\n", getpid());
      while (SDEBUG)
	;
    }

 /*
  * Initialise libgutenprint
  */

  stp_init();

  stp_set_output_codeset("UTF-8");

#ifdef INIT_I18N_UI
  INIT_I18N_UI();
#else
  /*
   * With GCC and glib 1.2, there will be a warning here about braces in
   * expressions.  Getting rid of it causes more problems than it solves.
   * In particular, turning on -ansi on the command line causes a number of
   * other useful things, such as strcasecmp, popen, and snprintf to go away
   */
  INIT_LOCALE (PACKAGE);
#endif

  stpui_printer_initialize(&gimp_vars);
  /*
   * Initialize parameter data...
   */

  run_mode = (GimpRunMode)param[0].data.d_int32;

  values = g_new (GimpParam, 1);

  values[0].type          = GIMP_PDB_STATUS;
  values[0].data.d_status = GIMP_PDB_SUCCESS;

  *nreturn_vals = 1;
  *return_vals  = values;

  image_ID = param[1].data.d_int32;
  drawable_ID = param[2].data.d_int32;

  image_filename = gimp_image_get_filename (image_ID);
  if (image_filename)
    {
      if (strchr(image_filename, '/'))
	image_filename = strrchr(image_filename, '/') + 1;
      stpui_set_image_filename(image_filename);
      /* g_free(image_filename); */
    }
  else
    stpui_set_image_filename("Untitled");

  /*  eventually export the image */
  switch (run_mode)
  {
  case GIMP_RUN_INTERACTIVE:
    case GIMP_RUN_WITH_LAST_VALS:
      gimp_ui_init ("print", TRUE);
      export = gimp_export_image (&image_ID, &drawable_ID, "Print",
                                  (GIMP_EXPORT_CAN_HANDLE_RGB |
                                   GIMP_EXPORT_CAN_HANDLE_GRAY |
                                   GIMP_EXPORT_CAN_HANDLE_INDEXED |
                                   GIMP_EXPORT_CAN_HANDLE_ALPHA));
      if (export == GIMP_EXPORT_CANCEL)
      {
        values[0].data.d_status = GIMP_PDB_EXECUTION_ERROR;
        fprintf(stderr,"Cannot handle image type\n");
        return;
      }
      break;
  default:
    break;
  }

  /*
   * Get drawable...
   */

  drawable = gimp_drawable_get (drawable_ID);
  stpui_set_image_dimensions(drawable->width, drawable->height);
  gimp_image_get_resolution (image_ID, &xres, &yres);
  stpui_set_image_resolution(xres, yres);
  stpui_set_image_channel_depth(8);
  base_type = gimp_image_base_type(image_ID);
  switch (base_type)
  {
  case GIMP_INDEXED:
  case GIMP_RGB:
    stpui_set_image_type("RGB");
    break;
  case GIMP_GRAY:
    stpui_set_image_type("Whitescale");
    break;
  default:
    break;
  }

  image = Image_GimpDrawable_new(drawable, image_ID);
  stp_set_float_parameter(gimp_vars.v, "AppGamma", gimp_gamma());

  /*
   * See how we will run
   */

  switch (run_mode)
  {
  case GIMP_RUN_INTERACTIVE:
    /*
     * Get information from the dialog...
     */

    if (!do_print_dialog (name, image_ID))
    {
      values[0].data.d_status = GIMP_PDB_EXECUTION_ERROR;
    }
    else
    {
      stpui_plist_copy(&gimp_vars, stpui_get_current_printer());
    }
    break;

  case GIMP_RUN_NONINTERACTIVE:
    /*
     * Make sure all the arguments are present...
     */
    if (nparams < 9)
      values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
    else
    {
      gimp_image_get_resolution(image_ID, &xres, &yres);
      gdouble pixwidth = gimp_drawable_width(drawable_ID);
      gdouble pixheight = gimp_drawable_height(drawable_ID);
      gdouble pointwidth = gimp_pixels_to_units(pixwidth, GIMP_UNIT_POINT, xres);
      gdouble pointheight = gimp_pixels_to_units(pixheight, GIMP_UNIT_POINT, yres);

      stp_set_height(gimp_vars.v, pointheight);
      stp_set_width(gimp_vars.v, pointwidth);

      /*
       * Avoid calling stpui_print with an invalid driver (SEGFAULT)
       */

      if (! stp_get_printer_by_driver(param[3].data.d_string) )
      {
        values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
        fprintf(stderr, "Unknown driver %s\n", stp_get_driver(gimp_vars.v));
        break;
      }

      stp_set_driver(gimp_vars.v, param[3].data.d_string);

      stpui_plist_set_queue_name(&gimp_vars, param[4].data.d_string);

      /*
       * Left offset (points, -1 = centered)
       */

      stp_set_left(gimp_vars.v, param[6].data.d_float);

      /*
       * Top offset (points, -1 = centered)
       */

      stp_set_top(gimp_vars.v, param[5].data.d_float);

      /*
       * Parse remaining parameters from key-value string array
       */

      int kv_arr_len = param[7].data.d_int32;

      if (kv_arr_len % 2 != 0)
      {
        /*
         * Key with no Value
         */

        values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
        fprintf(stderr,"Key with no value provided\n");
      } else {

	int k = 0;
        for( k=0; k<kv_arr_len && values[0].data.d_status == GIMP_PDB_SUCCESS; k+=2 )
        {
          char *key = param[8].data.d_stringarray[k];
          char *value = param[8].data.d_stringarray[k + 1];
          char *endptr = NULL;
          float float_value = 0;

          stp_describe_parameter(gimp_vars.v, key, &desc);

          switch(desc.p_type)
          {
          case STP_PARAMETER_TYPE_STRING_LIST:

            /*
             * Some useful string stp parameters
             *
             * "PrintingMode": BW, Color
             * "Resolution": "300", "720", etc.
             * "PageSize": "Letter", "A4", etc. TODO: Support Custom
             * "MediaType": "Plain", "Glossy", etc.
             * "InputSlot": "Tray1", "Manual", etc.
             * "ColorCorrection": Color Correction model
             * "InkType": Type of ink or cartridge
             * "InkSet": Set of inks to use
             * "DitherAlgorithm": Dither algorithm
             * "Weave": Weave method
             * "PrintingDirection": "Bidirectional", "Unidirectional"
             */

            stp_set_string_parameter(gimp_vars.v, key, value);
            break;

          case STP_PARAMETER_TYPE_INT:
            stp_set_int_parameter(gimp_vars.v, key, atoi(value));
            break;

          case STP_PARAMETER_TYPE_BOOLEAN:
            stp_set_boolean_parameter(gimp_vars.v, key, atoi(value));
            break;

          case STP_PARAMETER_TYPE_DOUBLE:
          case STP_PARAMETER_TYPE_DIMENSION:

            /*
             * Some useful floating point stp parameters
             *
             * "Brightness"  0-400%
             * "Gamma"  Output gamma 0.1 - 3.0
             * "Contrast" 0.1 - 3.0
             * "Saturation"  0-1000%
             * "Density"  0-200%
             * "DropSize1"  0.0-1.0
             * "DropSize2"  0.0-1.0
             * "DropSize3"  0.0-1.0
             */

            float_value = strtof(value, &endptr);
            if (float_value == 0 && endptr == value)
            {
              /*
               * No conversion was performed -- invalid floating point number
               */
              *nreturn_vals = 1;
              values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
              fprintf(stderr,"Invalid floating point value provided for key: %s\n", key);
            }
            else
            {
	      if (desc.p_type == STP_PARAMETER_TYPE_DOUBLE)
		stp_set_float_parameter(gimp_vars.v, key, float_value);
	      else
		stp_set_dimension_parameter(gimp_vars.v, key, float_value);
            }
            break;

          case STP_PARAMETER_TYPE_CURVE:
          case STP_PARAMETER_TYPE_FILE:
          case STP_PARAMETER_TYPE_RAW:
          case STP_PARAMETER_TYPE_ARRAY:
            values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
            fprintf(stderr,"Parameter type unsupported in gimp2 plugin for parameter %s\n", key);
            break;

          case STP_PARAMETER_TYPE_INVALID:

            /*
             * Output scaling (0-100%, -PPI)
             */

            if (strncmp("Scaling", key, 7) == 0)
            {
              float_value = strtof(value, &endptr);
              if (float_value == 0 && endptr == value)
              {
                /*
                 * No conversion was performed -- invalid floating point number
                 */

                values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
                fprintf(stderr,"Invalid floating point value provided for key: Scaling\n");
              }
              else
              {
                gimp_vars.scaling = float_value;

                if (gimp_vars.scaling == 0) {
                  values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
                  fprintf(stderr,"Scaling cannot be 0\n");
                }
                else if (gimp_vars.scaling > 0)
                {
                  /*
                   * Scaling > 0 in %
                   */

                  stp_set_width(gimp_vars.v, pointwidth * gimp_vars.scaling / 100.);
                  stp_set_height(gimp_vars.v, pointheight * gimp_vars.scaling / 100.);
                }
                else /* gimp_vars < 0 */
                {
                  /*
                   * Scaling < 0 in DPI
                   */

                  pointwidth = gimp_pixels_to_units(pixwidth, GIMP_UNIT_POINT, -gimp_vars.scaling);
                  pointheight = gimp_pixels_to_units(pixheight, GIMP_UNIT_POINT, -gimp_vars.scaling);
                  stp_set_width(gimp_vars.v, pointwidth);
                  stp_set_height(gimp_vars.v, pointheight);
                }
              }
              break;
            }

            /*
             * Output orientation (-1 = auto, 0 = portrait, 1 = landscape)
             */

            else if (strncmp("Orientation", key, 11) == 0)
            {
              gimp_vars.orientation = atoi(value);
              break;
            }

          default:
            values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
            fprintf(stderr,"Parameter unsupported in gimp2 plugin for parameter %s\n", key);
            break;
          }
	  stp_parameter_description_destroy(&desc);
        }
      }
    }
    break;

  case GIMP_RUN_WITH_LAST_VALS:
    values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
    break;

  default:
    values[0].data.d_status = GIMP_PDB_CALLING_ERROR;
    break;
  }

  if (gimp_thumbnail_data)
    g_free(gimp_thumbnail_data);

  /*
   * Print the image...
   */
  if (values[0].data.d_status == GIMP_PDB_SUCCESS)
  {
    /*
     * Set the tile cache size...
     */

    if (drawable->height > drawable->width)
      gimp_tile_cache_ntiles ((drawable->height + gimp_tile_width () - 1) /
                              gimp_tile_width () + 1);
    else
      gimp_tile_cache_ntiles ((drawable->width + gimp_tile_width () - 1) /
                              gimp_tile_width () + 1);

    if (! stpui_print(&gimp_vars, image))
    {
      values[0].data.d_status = GIMP_PDB_EXECUTION_ERROR;
    }

    /*
     * Store data...
     * FIXME! This is broken!
     */

#if 0
    if (run_mode == GIMP_RUN_INTERACTIVE)
      gimp_set_data (PLUG_IN_NAME, vars, sizeof (vars));
#endif
  }

  /*
   * Detach from the drawable...
   */
  gimp_drawable_detach (drawable);

  if (export == GIMP_EXPORT_EXPORT)
    gimp_image_delete (image_ID);
  stp_vars_destroy(gimp_vars.v);
}

/*
 * 'do_print_dialog()' - Pop up the print dialog...
 */

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat-nonliteral"
static void
gimp_errfunc(void *file, const char *buf, size_t bytes)
{
  char formatbuf[32];
  snprintf(formatbuf, 31, "%%%lus", (unsigned long) bytes);
  g_message(formatbuf, buf);
}
#pragma GCC diagnostic pop

static gint
do_print_dialog (const gchar *proc_name,
		 gint32       image_ID)
{
 /*
  * Generate the filename for the current user...
  */
  char *filename = gimp_personal_rc_file (cast_safe("printrc"));
  stpui_set_printrc_file(filename);
  g_free(filename);
  if (! getenv("STP_PRINT_MESSAGES_TO_STDERR"))
    stpui_set_errfunc(gimp_errfunc);
  stpui_set_thumbnail_func(stpui_get_thumbnail_data_function);
  stpui_set_thumbnail_data(gint2p(image_ID));
  return stpui_do_print_dialog();
}
